### ISSUE TITLE
Add a warning when collecting 'async def' functions

### ISSUE THREAD

**Issue by nicoddemus** [pytest-dev/pytest#2224]
As discussed in #2175, while it does not seem possible at the moment to support `asyncio` functions out of the box because it would be implemented in detriment of other async frameworks, we should at the very least emit a warning when collecting awaitable functions.

**Comment by pawelad** [comment]
How do you see that working with mentioned plugins like [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio) and [pytest-tornado](https://github.com/eugeniy/pytest-tornado)?

I think the correct behaviour would be to not display the warning if the test was successfully marked by them, but as fas as I understand (and please correct me if I'm wrong) it would mean we have to keep a list of 'allowed' markers/plugins.

I would be happy to help with the implementation but just want to make sure that I understand what is the expected behaviour.

**Comment by nicoddemus** [comment]
Unless someone has a better idea, I think just checking for the known marks would be sufficient. We should even add to the warning message a recommendation for the user to check out `pytest-asyncio` or `pytest-tornado` for async goodness, so having a list of known marks doesn't seem problematic to me.

**Comment by pawelad** [comment]
If it's good enough for you then it's good enough for me : )

But do you have in mind some kind of a 'vetting' process or would you rather just allow all async helping plugins, no matter how active they are? In other words, where do you draw the line? I'm fine with there not being a line, but then again it won't be a list of 'recommended' plugins but a list of 'known' ones. And please do tell me if I'm overthinking this : -)

That said - any other suggestions for the list? From my quick research I also found [pytest-tornasync](https://github.com/eukaryote/pytest-tornasync) and [pytest-aiohttp](https://github.com/aio-libs/pytest-aiohttp/) (although they use fixtures instead of markers, which may complicate the solution a little).

**Comment by nicoddemus** [comment]
> But do you have in mind some kind of a 'vetting' process or would you rather just allow all async helping plugins, no matter how active they are? In other words, where do you draw the line?

My initial reaction is to list all "known" plugins, except if one of them is marked as unmaintained by the author.

> From my quick research I also found pytest-tornasync and pytest-aiohttp (although they use fixtures instead of markers, which may complicate the solution a little).

Hmmm perhaps a solution would be to look at the list of installed plugins instead, and when one of the known async plugins is installed we just suppress all warnings related to async tests.

**Comment by RonnyPfannschmidt** [comment]
can we please assume people can at least google and just say that they need a pytest plugin fitting the async framework they use

**Comment by RonnyPfannschmidt** [comment]
also the async plugins should handle the async tests

**Comment by nicoddemus** [comment]
>can we please assume people can at least google and just say that they need a pytest plugin fitting the async framework they use

Fair enough.

> also the async plugins should handle the async tests

They already do, I'm not sure I get your point...

**Comment by RonnyPfannschmidt** [comment]
then they should not see a warning

**Comment by nicoddemus** [comment]
> then they should not see a warning

Do you mean that we should just emit a warning and assume that the plugins will override the appropriate hook (`pytest_runtest_call`) effectively silencing our warning?

**Comment by RonnyPfannschmidt** [comment]
correct

**Comment by nicoddemus** [comment]
Oh yeah you are right, no need for complicated checks then, which makes everything much easier.

@pawelad is this course of action clear or do you need some pointers?

Also, should this go into `master` or `features`? I believe it is harmless to add it to `master`, I don't see how it would break any test suite today (and if it breaks, it is because people are using async tests without a proper plugin installed).

**Comment by pawelad** [comment]
As far as I understand - no checking, no special conditions, just display the warning when collected test is an awaitable.

To be honest I don't really see how's that better, but then again I didn't go into code structure yet.

**Comment by pawelad** [comment]
I tried both the approach of the original PR (#2175) and plugging in the checking if function is a coroutine [here](https://github.com/pytest-dev/pytest/blob/master/_pytest/python.py#L179) but no matter what I seem to do I get this stacktrace:

```
$ pytest testing/python/collect.py
====================================================================================== test session starts ======================================================================================
platform darwin -- Python 3.5.3, pytest-3.0.8.dev, py-1.4.33, pluggy-0.4.0
rootdir: /Users/pawelad/Dev/Code/Forks/pytest, inifile: tox.ini
plugins: hypothesis-3.6.1
collected 77 items 

testing/python/collect.py .........................F...................................................

=========================================================================================== FAILURES ============================================================================================
___________________________________________________________________________ TestFunction.test_async_function_warning ____________________________________________________________________________

self = <collect.TestFunction object at 0x10c795ef0>
testdir = <Testdir local('/private/var/folders/01/qtz5hrt13cxgy__flq7zjgb00000gn/T/pytest-of-pawelad/pytest-40/testdir/test_async_function_warning0')>

    @pytest.mark.skipif(sys.version_info < (3, 4),
                        reason='async features available in Python 3.4+')
    def test_async_function_warning(self, testdir):
        with pytest.warns(RuntimeWarning) as record:
            testdir.makepyfile("""
                    import asyncio
                    import types
                    @types.coroutine
                    def test_coroutine():
                        yield from asyncio.sleep(1)
                """)
    
>           reprec = testdir.inline_run()
E           Failed: DID NOT WARN. No warnings of type (<class 'RuntimeWarning'>,) was emitted. The list of emitted warnings is: [].

/Users/pawelad/Dev/Code/Forks/pytest/testing/python/collect.py:403: Failed
------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------
============================= test session starts ==============================
platform darwin -- Python 3.5.3, pytest-3.0.8.dev, py-1.4.33, pluggy-0.4.0
rootdir: /private/var/folders/01/qtz5hrt13cxgy__flq7zjgb00000gn/T/pytest-of-pawelad/pytest-40/testdir/test_async_function_warning0, inifile:
plugins: hypothesis-3.6.1
collected 0 items / 1 errors

==================================== ERRORS ====================================
_______________ ERROR collecting test_async_function_warning.py ________________

self = <CallInfo when='memocollect' exception: <function test_coroutine at 0x10c655d08> yielded non callable test <Future pending>>
func = <bound method Collector._memocollect of <Generator 'test_coroutine'>>
when = 'memocollect'

    def __init__(self, func, when):
        #: context of invocation: one of "setup", "call",
        #: "teardown", "memocollect"
        self.when = when
        self.start = time()
        try:
>           self.result = func()

/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/runner.py:163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/main.py:476: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/main.py:347: in _memoizedcall
    res = function()
/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/main.py:476: in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Generator 'test_coroutine'>

    def collect(self):
        # test generators are seen as collectors but they also
        # invoke setup/teardown on popular request
        # (induced by the common "test_*" naming shared with normal tests)
        from _pytest import deprecated
        self.session._setupstate.prepare(self)
        # see FunctionMixin.setup and test_setupstate_is_preserved_134
        self._preservedparent = self.parent.obj
        l = []
        seen = {}
        for i, x in enumerate(self.obj()):
            name, call, args = self.getcallargs(x)
            if not callable(call):
>               raise TypeError("%r yielded non callable test %r" %(self.obj, call,))
E               TypeError: <function test_coroutine at 0x10c655d08> yielded non callable test <Future pending>

/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/python.py:628: TypeError
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.28 seconds ============================
============================================================================== 1 failed, 76 passed in 6.47 seconds ==============================================================================
```

My test:
```python
    @pytest.mark.skipif(sys.version_info < (3, 4),
                        reason='async features available in Python 3.4+')
    def test_async_function_warning(self, testdir):
        with pytest.warns(RuntimeWarning) as record:
            testdir.makepyfile("""
                import asyncio
                import types
                @types.coroutine
                def test_coroutine():
                    yield from asyncio.sleep(1)
            """)

            reprec = testdir.inline_run()
            # reprec.assertoutcome(skipped=1)

        assert len(record) == 1
        assert record[0].message.args[0] == (
            "pytest doesn't currently support coroutines "
            "testing out of the box"
        )
```

Any idea what could be the reason?

**Comment by pawelad** [comment]
Changed code in the hooks in case that helps:

```python
@pytest.hookimpl(trylast=True)
def pytest_pyfunc_call(pyfuncitem):
    testfunction = pyfuncitem.obj
    if iscoroutinefunction(testfunction):
        warnings.warn(
            "pytest doesn't currently support coroutines "
            "testing out of the box", RuntimeWarning
        )
        return False

    if pyfuncitem._isyieldedfunction():
        testfunction(*pyfuncitem._args)
    else:
        funcargs = pyfuncitem.funcargs
        testargs = {}
        for arg in pyfuncitem._fixtureinfo.argnames:
            testargs[arg] = funcargs[arg]
        testfunction(**testargs)
    return True

@pytest.hookimpl(hookwrapper=True)
def pytest_pycollect_makeitem(collector, name, obj):
    outcome = yield
    res = outcome.get_result()
    if res is not None:
        return
    # nothing was collected elsewhere, let's do it here
    if isclass(obj):
        if collector.istestclass(obj, name):
            Class = collector._getcustomclass("Class")
            outcome.force_result(Class(name, parent=collector))
    elif collector.istestfunction(obj, name):
        # mock seems to store unbound methods (issue473), normalize it
        obj = getattr(obj, "__func__", obj)

        # Check if the function is in fact a coroutine
        if iscoroutinefunction(obj):
            collector.warn(
                code="C3",
                message="pytest doesn't currently support coroutines "
                        "testing out of the box",
            )
        # We need to try and unwrap the function if it's a functools.partial
        # or a funtools.wrapped.
        # We musn't if it's been wrapped with mock.patch (python 2 only)
        elif not (isfunction(obj) or isfunction(get_real_func(obj))):
            collector.warn(code="C2", message=
                "cannot collect %r because it is not a function."
                % name, )
        elif getattr(obj, "__test__", True):
            if is_generator(obj):
                res = Generator(name, parent=collector)
            else:
                res = list(collector._genfunctions(name, obj))
            outcome.force_result(res)
```

**Comment by maiksensi** [comment]
Is this still an issue / a todo? Anyone know the current status of this?

**Comment by RonnyPfannschmidt** [comment]
still a todo

**Comment by aparamon** [comment]
Will the warnings be emitted for `async def` tests with https://github.com/eukaryote/pytest-tornasync installed?

**Comment by nicoddemus** [comment]
Will be available in pytest 4.4.

**Comment by uriva** [comment]
Stumbled across this warning today: `async def functions are not natively supported and have been skipped`. But didn't get a test name / filename, so had to look for it >40 files. Is there a better way?
